spa/src/pages/DeferredInstallation/index.tsx

```diff
@@ -0,0 +1,211 @@
+`;
+const linkStyle = css`
+	cursor: pointer;
+	padding-left: 0;
+	padding-right: 0;
+`;
+
+
+const DeferredInstallation = () => {
+	const [searchParams] = useSearchParams();
+	const requestId = searchParams.get("requestId") || "";
+	const navigate = useNavigate();
+	const username = OAuthManager.getUserDetails().username || "";
+
+	const [isPopupBlocked, setPopupBlocked] = useState<boolean>(false);
+	const onPopupBlocked = () => setPopupBlocked(true);
+
+	const [isLoading, setIsLoading] = useState(false);
+	const [isLoggedIn, setIsLoggedIn] = useState(false);
+	const [error, setError] = useState<ErrorObjType | undefined>(undefined);
+
+	const [jiraHost, setJiraHost] = useState("");
+	const [orgName, setOrgName] = useState("");
+
+	const [forbidden, setForbidden] = useState(false);
+
+	// Extract the info from the requestId
+	useEffect(() => {
+		const extractFromRequestId = async () => {
+			const extractedPayload = await DeferralManager.extractFromRequestId(requestId);
+			if (extractedPayload instanceof AxiosError) {
+				setError(modifyError(
+					{ errorCode: "INVALID_DEFERRAL_REQUEST_ID"},
+					{},
+					{ onClearGitHubToken: () => {}, onRelogin: () => {}, onPopupBlocked }
+				));
+				setIsLoading(true);
+			} else {
+				setJiraHost(extractedPayload.jiraHost as string);
+				setOrgName(extractedPayload.orgName);
+			}
+		};
+		extractFromRequestId();
+	}, [ requestId ]);
+
+	// Finish the OAuth dance if authenticated
+	useEffect(() => {
+		const handler = async (event: MessageEvent) => {
+			if (event.data?.type === "oauth-callback" && event.data?.code) {
+				const response: boolean | AxiosError = await OAuthManager.finishOAuthFlow(event.data?.code, event.data?.state);
+				setIsLoading(false);
+				if (response instanceof AxiosError) {
+					return;
+				}
+				setIsLoggedIn(true);
+			}
+		};
+		window.addEventListener("message", handler);
+		return () => {
+			window.removeEventListener("message", handler);
+		};
+	}, []);
+
+	// Set the token/username after authentication
+	useEffect(() => {
+		const connectDeferredOrgOrg = async () => {
+			if (requestId) {
+				setIsLoading(true);
+				const status: boolean | AxiosError = await DeferralManager.connectOrgByDeferral(requestId);
+				if (status instanceof AxiosError) {
+					setForbidden(true);
+				}
+				else if (status) {
+					console.log("Successfully connected now navigate");
+					setForbidden(false);
+					navigate("/spa/connected",{ state: { orgLogin: orgName, connectedByDeferral: true } });
+				} else {
+					setForbidden(true);
+				}
+				setIsLoading(false);
+			}
+		};
+
+		isLoggedIn && connectDeferredOrgOrg();
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ isLoggedIn ]);
+
+	// Authenticate if no token/username is set
+	const authenticate = async () => {
+		setIsLoading(true);
+		try {
+			await OAuthManager.authenticateInGitHub({
+				onWinClosed: () => {
+					setIsLoading(false);
+					console.log("Successfully authenticated", username);
```

Comment:
Shouldn't we call `connectDeferredOrgOrg()` here? So that we don't have to do another `useEffect`?

Reply:
The `connectDeferredOrg` should happen after the whole OAuth flow is finished, which happens on the other useEffect with `OAuthManager.finishOAuthFlow`.
And I've kept it in separate useEffect for better readability and avoiding massive code in a single useEffect.